xbmc/filesystem/FileCache.cpp

```diff
@@ -398,11 +395,9 @@ void CFileCache::Process()
     // avoid uncertainty at start of caching
     m_writeRateActual = average.Rate(m_writePos, 1000);
 
-    // Update forward cache size
-    m_forward = m_pCache->WaitForData(0, 0);
-
     // NOTE: Hysteresis (20-80%) for filling-logic
-    const float level = (m_forwardCacheSize == 0) ? 0.0 : (float) m_forward / m_forwardCacheSize;
+    const int64_t forward = m_pCache->WaitForData(0, 0);
+    const float level = (m_forwardCacheSize == 0) ? 0.0 : (float) forward / m_forwardCacheSize;
     if (level > 0.8f)
```

Comment:
Is there a possibilitiy that cache size == m_chunksize? Then this would only be true if entirely filled and be false else? Do we need to take care that at least 2,3 or 4 times the chunksize fits into the cache?

Reply:
Are you referring to the line for level? Since I'm not sure what you mean here. Could you elaborate?

Reply:
Yes. 0.8 means 80%.

If you have a buffer that is just as large as your chunk size. This will only be true if it's full and false if it's empty.

We have a similar "issue" in our Audio sinks. There is a buffer (cache buffer) and there is a period (your chunksize). We make sure to have at least 4 periods in the buffer to not run dry and not to run too much on CPU.

Reply:
For further explain my question: If you only allow chunksize into the buffer and the buffer is chunksize large, then it has to be entirely empty, so that you can add stuff again. This needs to be taken into account.

Reply:
Yes, but when file size = chunk size = cache size, it no longer works as a fifo but as a linear buffer. So I don't think we have a problem here.

Reply:
Btw. please note, like I said before, caching such small files is nonsense and adds more overhead than benefit afaik.

Reply:
What I have written is also true for any size of files. This is why I asked - if you say, that is not a problem, we can continue here.

Reply:
Ah yes, now I get it . Good point. I think we need a std:min/max with chunk size here for the 20/80 check

Reply:
Let's try to have the cache at a minimum 1.5 times chunksize large?

Reply:
Or if we want to keep 20 and 80 and we want to do something when we hit 80. Then 20% should match m_chunksize and we need 5 times m_chunksize of cache.

if we would compute all in chunksizes. Then let's do 4 times m_chunksize and check on 25% and 75%.

Reply:
First of all, keep in mind this logic is (currently) only used by VideoPlayer which practically means it's never used for tiny files. But assume a file of let's say 2MiB with a source chunk size of 1MiB, making the cache size 4 or 5 MiB won't make any difference since the cache never fully fills. In that case it's also not that important that this logic does anything since the cache doesn't operate as a FIFO for such a small file, so there aren't any issues with dry-running and such. Note that the 20% and 80% were just sort of randomly chosen back then since I wasn't sure what the most optimal values were. The bigger the difference between the 2 percentages, the more reliable this logic works.

The one thing, I think, would be sane to do is to make sure the cache size is always at least chunk size (or perhaps even 2*, not sure) capped by file size. However, this is unrelated to the level logic here.

Reply:
Your proposal at the end (2 times chunksize at least, or chunksize if filesize < chunksize) sounds sane to me.

Reply:
I'd opt for 2*chunksize as well